### PR TITLE
Handle long messages in event parser

### PR DESCRIPTION
**Pull request by aldld**
This PR fixes a bug that causes the `go-sse` client to drop messages that are longer than 4096 bytes with an Unexpected EOF error.

The issue here is that the `Scanner` that gets used to fill the input buffer for the byte parser (which gets used to extract fields) may not call the `SplitFunc` with a buffer containing a full token. In that case, the split function needs to request more data, signalling to the `Scanner` to call the `SplitFunc ` again with a larger buffer.

This PR includes a new test case in `TestReaderParser` that fails if this change is not present, along with a new test in `internal/parser/split_func_test.go`.

### CODE REVIEW

File: internal/parser/reader_parser.go [hunk 1]
```diff
@@ -22,6 +22,12 @@ var splitFunc bufio.SplitFunc = func(data []byte, _ bool) (advance int, token []
...
 
+	if advance == len(data) && !atEOF {
+		// We have reached the end of the buffer but have not yet seen two consecutive
+		// newline sequences, so we request more data.
+		return 0, nil, nil
+	}
+
 	if l := len(data); advance < l {
 		advance++
 		if advance < l && data[advance-1] == '\r' && data[advance] == '\n' {
```
Comment (tmaxmax) [inline internal/parser/reader_parser.go:33]: Can we write it like this maybe, so we don't repeat the `len(data)`? It's also a good opportunity to document some surrounding code.
```suggestion
	if l := len(data); advance == l && !atEOF {
		// We have reached the end of the buffer but have not yet seen two consecutive
		// newline sequences, so we request more data.
		return 0, nil, nil
	} else if advance < l {
		// We have found a newline. Consume the end-of-line sequence.
		advance++
		// Consume one more character if end-of-line is \r\n.
		if advance < l && data[advance-1] == '\r' && data[advance] == '\n' {
```

File: internal/parser/reader_parser.go [hunk 2]
```diff
@@ -5,7 +5,7 @@ import (
 	"io"
 )
 
-var splitFunc bufio.SplitFunc = func(data []byte, _ bool) (advance int, token []byte, err error) {
+var splitFunc bufio.SplitFunc = func(data []byte, atEOF bool) (advance int, token []byte, err error) {
 	if len(data) == 0 {
 		return
```
Comment (tmaxmax) [inline internal/parser/reader_parser.go:10]: Also, while we're here, can we get rid of naked returns? There is also one at the end of the function, which should be `return advance, token, nil`.
```suggestion
		return 0, nil, nil
```

**Review by tmaxmax — CHANGES_REQUESTED** [review]
Thank you for raising the issue and the PR! The requested changes address only some style preferences. It's also great you've added tests for this edge case – seems like I missed that part of `bufio.Scanner`'s behavior. I appreciate you taking time to find and fix this issue. Was it easy navigating the codebase? If you have any naming/layout/other suggestions, please let me know.

**Comment by tmaxmax** [comment]
Also make sure to rebase on master – I've updated and fixed the actions.

File: internal/parser/reader_parser.go (hunk 2 above)
Comment (aldld) [reply internal/parser/reader_parser.go:10]: Agreed, will remove the naked returns from this function

File: internal/parser/reader_parser.go (hunk 1 above)
Comment (aldld) [reply internal/parser/reader_parser.go]: Done. I also added a comment at the top of `splitFunc` to try and describe what it does at a high level, since I had a bit of a hard time figuring it out the first time around.

**Comment by aldld** [comment]
> Thank you for raising the issue and the PR! The requested changes address only some style preferences. It's also great you've added tests for this edge case – seems like I missed that part of `bufio.Scanner`'s behavior. I appreciate you taking time to find and fix this issue. Was it easy navigating the codebase? If you have any naming/layout/other suggestions, please let me know.

Thanks for the feedback! I've rebased and updated the PR based on your comments. I found most of the codebase quite easy to follow, particularly the top-level package and the readme documentation. It did take me some time to wrap my head around the `internal/parser` package, mainly in terms of understanding the relationship between the different functions named `Scan` that get called by `ReaderParser.Scan()`. It was a bit difficult to keep in my head the difference between e.g. `r.p.Scan()` and `r.sc.Scan()`, so perhaps naming those fields a bit differently (maybe `eventScanner` and `fieldScanner`?) would help clarify what's happening at each step.

**Comment by tmaxmax** [comment]
> > Thank you for raising the issue and the PR! The requested changes address only some style preferences. It's also great you've added tests for this edge case – seems like I missed that part of `bufio.Scanner`'s behavior. I appreciate you taking time to find and fix this issue. Was it easy navigating the codebase? If you have any naming/layout/other suggestions, please let me know.
> 
> Thanks for the feedback! I've rebased and updated the PR based on your comments. I found most of the codebase quite easy to follow, particularly the top-level package and the readme documentation.
> 
> It did take me some time to wrap my head around the `internal/parser` package, mainly in terms of understanding the relationship between the different functions named `Scan` that get called by `ReaderParser.Scan()`. It was a bit difficult to keep in my head the difference between e.g. `r.p.Scan()` and `r.sc.Scan()`, so perhaps naming those fields a bit differently (maybe `eventScanner` and `fieldScanner`?) would help clarify what's happening at each step.
> 
> Other than that, though, I had no trouble understanding and navigating the codebase.

That's a very useful suggestion, thank you! I'll implement it in another PR. Will merge your PR and release a new version soon, as I've found another bug and an opportunity to remove some code from the codebase.

**Review by tmaxmax — APPROVED** [review]

**Comment by tmaxmax** [comment]
See [latest release](https://github.com/tmaxmax/go-sse/releases/tag/v0.4.3).
